zstash/hpss.py

```diff
@@ -95,7 +98,28 @@ def hpss_transfer(
             # Transfer file using `hsi`
             command: str = 'hsi -q "cd {}; {} {}"'.format(hpss, transfer_command, name)
             error_str: str = "Transferring file {} HPSS: {}".format(transfer_word, name)
-            run_command(command, error_str)
+            tries = retries + 1
+            while tries > 0:
+                tries -= 1
+                try:
+                    run_command(command, error_str)
+                    # Command was successful. No need to retry.
+                    break
+                except RuntimeError as e:
+                    actual_size = os.path.getsize(name)
+                    if (
+                        (tries > 0)
+                        and (transfer_type == "get")
+                        and cur
+                        and tars_table_exists(cur)
+                    ):
+                        cur.execute(f"select size from tars where name is '{name}';")
+                        expected_size: int = cur.fetchall()[0][0]
+                        if expected_size > actual_size:
+                            # Rerun since tar was incomplete
+                            # Break out of the for-loop and run the try-except block again
```

Comment:
Note: this comment is from an earlier iteration of work; this should be changed to simply "Run the try-except block again"

Reply:
@golaz re: `check the tar file size regardless of hsi get status` -- Sure, I can copy the `if expected_size > actual_size:` logic to the `try` block so that it is checked on both success and failure.

You also mentioned the logic should be at https://github.com/E3SM-Project/zstash/blob/main/zstash/extract.py#L435 (`if not os.path.exists(tfname):`) instead, but I'm not seeing what effect that has. If there's no `tfname`, then we have nothing to get a size from -- so I don't see what logic we would apply at this level. As for the case where `tfname` does exist, we have: `tfname` -> `file_path` in `hpss_get` -> `file_path` in `hpss_transfer` -> `path, name = os.path.split(file_path)` -> `actual_size = os.path.getsize(name)`.

Reply:
@forsyth2, regarding the check around https://github.com/E3SM-Project/zstash/blob/753452869a27e683b515fc285a5869fa304fdc40/zstash/extract.py#L435

I would modify to do the following: retrieve the tar from from HPSS 
- if it is not in the disk cache (what we do now)
- OR if the file is on disk cache, but its size does not match the entry in the database (assuming version of zstash archive has that information).

This would help recover from previous errors.